[X86] Generate `kmov` for masking integers

abhishek-kaushik22 opened the pull request:
When we have an integer used as a bit mask the llvm ir looks something like this
```
%1 = and <16 x i32> %.splat, <i32 1, i32 2, i32 4, i32 8, i32 16, i32 32, i32 64, i32 128, i32 256, i32 512, i32 1024, i32 2048, i32 4096, i32 8192, i32 16384, i32 32768>
%cmp1 = icmp ne <16 x i32> %1, zeroinitializer
```
where `.splat` is vector containing the mask in all lanes. The assembly generated for this looks like
```
vpbroadcastd    %ecx, %zmm0
vptestmd        .LCPI0_0(%rip), %zmm0, %k1
```
where we have a constant table of powers of 2.
Instead of doing this we could just move the relevant bits directly to `k` registers using a `kmov` instruction. 
```
kmovw   %ecx, %k1
```
This is faster and also reduces code size.

Commit:
Generate `kmov` for masking integers

When we have an integer used as a bit mask the llvm ir looks something like this
```
%1 = and <16 x i32> %.splat, <i32 1, i32 2, i32 4, i32 8, i32 16, i32 32, i32 64, i32 128, i32 256, i32 512, i32 1024, i32 2048, i32 4096, i32 8192, i32 16384, i32 32768>
  %cmp1 = icmp ne <16 x i32> %1, zeroinitializer
```
where `.splat` is vector containing the mask in all lanes.
The assembly generated for this looks like
```
vpbroadcastd    %ecx, %zmm0
vptestmd        .LCPI0_0(%rip), %zmm0, %k1
```
where we have a constant table of powers of 2.
Instead of doing this we could just move the relevant bits directly to `k` registers using a `kmov` instruction. This is faster and also reduces code size.

Commit:
Merge branch 'main' into kmov

abhishek-kaushik22 commented:
@e-kud @phoebewang @RKSimon @topperc can you please review?

phoebewang commented on llvm/test/CodeGen/X86/pr78897.ll:
```diff
@@ -256,8 +256,8 @@ define <16 x i8> @produceShuffleVectorForByte(i8 zeroext %0) nounwind {
 ; X64-AVX512:       # %bb.0: # %entry
-; X64-AVX512-NEXT:    vpbroadcastb %edi, %xmm0
-; X64-AVX512-NEXT:    vptestnmb {{\.?LCPI[0-9]+_[0-9]+}}(%rip), %xmm0, %k1
+; X64-AVX512-NEXT:    kmovw %edi, %k0
+; X64-AVX512-NEXT:    knotw %k0, %k1
```
Please check why 32-bit not affected?

abhishek-kaushik22 replied:
Sorry about that, 32-bit Selection DAG used `VBROADCAST_LOAD` and not `VBROADCAST` which is why it got skipped.

phoebewang commented on llvm/lib/Target/X86/X86ISelDAGToDAG.cpp:
```diff
@@ -4975,12 +4975,70 @@ bool X86DAGToDAGISel::tryVPTESTM(SDNode *Root, SDValue Setcc,
   if (CanFoldLoads) {
+    if (canUseKMOV()) {
+      auto Op = Src0.getOperand(0);
+      if (Op.getSimpleValueType() == MVT::i8) {
+        Op = SDValue(CurDAG->getNode(ISD::ZERO_EXTEND, dl, MVT::i32, Op));
+      }
```
Remove parentheses

phoebewang commented on llvm/lib/Target/X86/X86ISelDAGToDAG.cpp:
```diff
@@ -4975,12 +4975,70 @@ bool X86DAGToDAGISel::tryVPTESTM(SDNode *Root, SDValue Setcc,
+        return i != 0;
+      }
+
+      if (Element->getUniqueInteger() != k) {
+        return false;
+      }
```
Remove parentheses

phoebewang commented on llvm/lib/Target/X86/X86ISelDAGToDAG.cpp:
```diff
@@ -4975,12 +4975,70 @@ bool X86DAGToDAGISel::tryVPTESTM(SDNode *Root, SDValue Setcc,
+    if (!ConstVecType)
+      return false;
+
+    for (unsigned i = 0, e = ConstVecType->getNumElements(), k = 1; i != e;
```
Use upper case I, J, E.

phoebewang commented on llvm/lib/Target/X86/X86ISelDAGToDAG.cpp:
```diff
@@ -4975,12 +4975,70 @@ bool X86DAGToDAGISel::tryVPTESTM(SDNode *Root, SDValue Setcc,
+        return i != 0;
+      }
+
+      if (Element->getUniqueInteger() != k) {
```
`k` is meaningless, use a meaningful name or simply `1 << i` here.

Commit:
Review Changes

Commit:
Merge branch 'main' into kmov

RKSimon commented on llvm/test/CodeGen/X86/kmov.ll:
```diff
@@ -0,0 +1,205 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc < %s -mtriple=x86_64-unknown-unknown -mcpu=skylake-avx512 | FileCheck %s
+
+define dso_local void @foo_16_ne(ptr nocapture noundef writeonly %c, ptr nocapture noundef readonly %a, ptr nocapture noundef readonly %b, i32 noundef %mask) {
```
cleanup attributes - no need for dso_local / nocapture / noundef / writeonly in this context

RKSimon reviewed (commented):
Tests need some cleanup/simplification - ideally its good to precommit the tests with current codegen and then show the codegen change when the combine is added in a later commit.

What was the reason behind performing this in DAGToDAG and not as a regular DAG combine?

abhishek-kaushik22 commented:
> Tests need some cleanup/simplification - ideally its good to precommit the tests with current codegen and then show the codegen change when the combine is added in a later commit.

Thanks, I'll take care next time.

> What was the reason behind performing this in DAGToDAG and not as a regular DAG combine?

I wasn't sure what this could be combined to, I didn't find any useful patterns that gets selected as KMOV. I only see these patterns in tablegen.

```
/*544207*/    OPC_MorphNodeTo1None, TARGET_VAL(X86::KMOVWkr),
                  /*MVT::v16i1*/22, 1/*#Ops*/, 5,
              // Src: (insert_subvector:{ *:[v16i1] } immAllZerosV:{ *:[v16i1] }, (scalar_to_vector:{ *:[v1i1] } GR8:{ *:[i8] }:$src), 0:{ *:[iPTR] }) - Complexity = 15
              // Dst: (KMOVWkr:{ *:[v16i1] } (AND32ri:{ *:[i32] }:{ *:[i32] } (INSERT_SUBREG:{ *:[i32] } (IMPLICIT_DEF:{ *:[i32] }), GR8:{ *:[i8] }:$src, sub_8bit:{ *:[i32] }), 1:{ *:[i32] }))
```
If there's a simpler pattern that I missed, please let me know.

Commit:
Update tests

- Remove attributes
- Remove fast math flags
- Simplify tests by removing mask/loads

abhishek-kaushik22 commented on llvm/test/CodeGen/X86/kmov.ll:
```diff
@@ -0,0 +1,205 @@
+  %2 = tail call <16 x float> @llvm.masked.load.v16f32.p0(ptr %b, i32 4, <16 x i1> %hir.cmp.45, <16 x float> poison)
+  %3 = tail call <16 x float> @llvm.masked.load.v16f32.p0(ptr %a, i32 4, <16 x i1> %hir.cmp.45, <16 x float> poison)
+  %4 = fadd reassoc nsz arcp contract afn <16 x float> %2, %3
+  tail call void @llvm.masked.store.v16f32.p0(<16 x float> %4, ptr %c, i32 4, <16 x i1> %hir.cmp.45)
```
I've removed the masked load/stores by directly returning the vector, but it didn't really work for vector length 4 where we had a `X86ISD::PCMPEQ` node which isn't part of this pattern.

abhishek-kaushik22 commented on llvm/test/CodeGen/X86/kmov.ll:
```diff
@@ -0,0 +1,205 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc < %s -mtriple=x86_64-unknown-unknown -mcpu=skylake-avx512 | FileCheck %s
+
+define dso_local void @foo_16_ne(ptr nocapture noundef writeonly %c, ptr nocapture noundef readonly %a, ptr nocapture noundef readonly %b, i32 noundef %mask) {
```
I couldn't think of one, I have replaced foo with the pr number.

RKSimon commented:
This is what you're matching:
```
Type-legalized selection DAG: %bb.0 'pr120593_16_ne:'
SelectionDAG has 31 nodes:
  t0: ch,glue = EntryToken
  t2: i32,ch = CopyFromReg t0, Register:i32 %0
          t34: v16i32 = BUILD_VECTOR t2, t2, t2, t2, t2, t2, t2, t2, t2, t2, t2, t2, t2, t2, t2, t2
          t23: v16i32 = BUILD_VECTOR Constant:i32<1>, Constant:i32<2>, Constant:i32<4>, Constant:i32<8>, Constant:i32<16>, Constant:i32<32>, Constant:i32<64>, Constant:i32<128>, Constant:i32<256>, Constant:i32<512>, Constant:i32<1024>, Constant:i32<2048>, Constant:i32<4096>, Constant:i32<8192>, Constant:i32<16384>, Constant:i32<32768>
        t24: v16i32 = and t34, t23
        t26: v16i32 = BUILD_VECTOR Constant:i32<0>, Constant:i32<0>, Constant:i32<0>, Constant:i32<0>, Constant:i32<0>, Constant:i32<0>, Constant:i32<0>, Constant:i32<0>, Constant:i32<0>, Constant:i32<0>, Constant:i32<0>, Constant:i32<0>, Constant:i32<0>, Constant:i32<0>, Constant:i32<0>, Constant:i32<0>
      t28: v16i1 = setcc t24, t26, setne:ch
    t29: v16i8 = any_extend t28
  t32: ch,glue = CopyToReg t0, Register:v16i8 $xmm0, t29
  t33: ch = X86ISD::RET_GLUE t32, TargetConstant:i32<0>, Register:v16i8 $xmm0, t32:1
```
You'd probably be better off trying to fold this inside combineSETCC - there's already a number of similar patterns that get folded in there.

What you're after is folding this to something like: `(sext v16i8 (v16i1 bitcast(i16 trunc (i32 t2))))`

Commit:
Combine to KMOV

Combine to KMOV instead of doing it in ISEL

abhishek-kaushik22 commented:
> This is what you're matching:
> 
> ```
> Type-legalized selection DAG: %bb.0 'pr120593_16_ne:'
> SelectionDAG has 31 nodes:
>   t0: ch,glue = EntryToken
>   t2: i32,ch = CopyFromReg t0, Register:i32 %0
>           t34: v16i32 = BUILD_VECTOR t2, t2, t2, t2, t2, t2, t2, t2, t2, t2, t2, t2, t2, t2, t2, t2
>           t23: v16i32 = BUILD_VECTOR Constant:i32<1>, Constant:i32<2>, Constant:i32<4>, Constant:i32<8>, Constant:i32<16>, Constant:i32<32>, Constant:i32<64>, Constant:i32<128>, Constant:i32<256>, Constant:i32<512>, Constant:i32<1024>, Constant:i32<2048>, Constant:i32<4096>, Constant:i32<8192>, Constant:i32<16384>, Constant:i32<32768>
>         t24: v16i32 = and t34, t23
>         t26: v16i32 = BUILD_VECTOR Constant:i32<0>, Constant:i32<0>, Constant:i32<0>, Constant:i32<0>, Constant:i32<0>, Constant:i32<0>, Constant:i32<0>, Constant:i32<0>, Constant:i32<0>, Constant:i32<0>, Constant:i32<0>, Constant:i32<0>, Constant:i32<0>, Constant:i32<0>, Constant:i32<0>, Constant:i32<0>
>       t28: v16i1 = setcc t24, t26, setne:ch
>     t29: v16i8 = any_extend t28
>   t32: ch,glue = CopyToReg t0, Register:v16i8 $xmm0, t29
>   t33: ch = X86ISD::RET_GLUE t32, TargetConstant:i32<0>, Register:v16i8 $xmm0, t32:1
> ```
> 
> You'd probably be better off trying to fold this inside combineSETCC - there's already a number of similar patterns that get folded in there.
> 
> What you're after is folding this to something like: `(sext v16i8 (v16i1 bitcast(i16 trunc (i32 t2))))`

Thanks! I've changed to match this pattern as a DAG combine.

Commit:
Merge branch 'main' into kmov

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -55293,6 +55293,95 @@ static SDValue truncateAVX512SetCCNoBWI(EVT VT, EVT OpVT, SDValue LHS,
+    return SDValue();
+
+  SDValue Load = Op0.getOperand(1);
+  if (Load.getOpcode() != ISD::LOAD)
```
You should be able to use getTargetConstantBitsFromNode to extract the element bits/undefs for the load:
```cpp
APInt UndefElts;
SmallVector<APInt> EltBits;
if (!getTargetConstantBitsFromNode(Op0.getOperand(1), VT.getScalarSizeInBits(), UndefElts, EltBits))
  return SDValue();
```

Commit:
Use getTargetConstantBitsFromNode

Commit:
Merge branch 'main' into kmov

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -55311,48 +55311,35 @@ static SDValue combineAVX512SetCCToKMOV(EVT VT, SDValue Op0, ISD::CondCode CC,
+  unsigned Len = UndefElts.getBitWidth();
+  for (unsigned I = 1; I != Len; ++I) {
+    if (UndefElts[I]) {
+      for (unsigned J = I + 1; J != Len; ++J)
```
Can you use something like "UndefElts.extractBits(Len -(I + 1), I + 1).isAllOnes()" or "UndefElts.lshr(I + 1).isAllOnes()" ?

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -55447,6 +55447,82 @@ static SDValue truncateAVX512SetCCNoBWI(EVT VT, EVT OpVT, SDValue LHS,
+    unsigned Mask = (1ULL << Len) - 1;
+    SDValue ShiftedValue = DAG.getNode(ISD::SRL, DL, BroadcastOpVT, BroadcastOp,
+                                       DAG.getConstant(N, DL, BroadcastOpVT));
+    Masked = DAG.getNode(ISD::AND, DL, BroadcastOpVT, ShiftedValue,
```
Is this path tested? In all the tests below the pow2 masks appear to start at 1

abhishek-kaushik22 replied:
Sorry I missed that, added the test for this case as well.

Commit:
Update test

Commit:
Update X86ISelLowering.cpp

Commit:
Merge branch 'main' into kmov

Commit:
Update X86ISelLowering.cpp

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -55447,6 +55447,81 @@ static SDValue truncateAVX512SetCCNoBWI(EVT VT, EVT OpVT, SDValue LHS,
+    Masked = DAG.getNode(ISD::AND, DL, BroadcastOpVT, ShiftedValue,
+                         DAG.getConstant(Mask, DL, BroadcastOpVT));
+  }
+  SDValue Trunc = DAG.getNode(BroadcastOpVT.bitsGT(MVT::i16) ? ISD::TRUNCATE
```
DAG.getAnyExtOrTrunc

RKSimon commented on llvm/test/CodeGen/X86/kmov.ll:
```diff
@@ -0,0 +1,150 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc < %s -mtriple=x86_64-- -mcpu=x86-64-v4 | FileCheck %s
+
+define <16 x i1> @mask_16(i32 %mask) {
```
All these tests are for i32 masks - we need a mixture of i8/i16/i32/i64 tests - preferably across 128/256/512 vectors as well - not need for full duplication, but you do need to make sure you testing all possibly legal types properly.

abhishek-kaushik22 replied:
I've added more tests, but this pattern didn't work for 256-bit vectors such as (`v16i16`, `v8i32`, `v4i64`) because a `X86ISD::PCMPEQ` node is generated before we can combine the `setcc`.
I can try and fix this after this gets merged.

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -55447,6 +55447,81 @@ static SDValue truncateAVX512SetCCNoBWI(EVT VT, EVT OpVT, SDValue LHS,
   return SDValue();
 }
 
+static SDValue combineAVX512SetCCToKMOV(EVT VT, SDValue Op0, ISD::CondCode CC,
```
A brief description would be nice :)

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -55447,6 +55447,81 @@ static SDValue truncateAVX512SetCCNoBWI(EVT VT, EVT OpVT, SDValue LHS,
+            DAG.getConstant(APInt::getAllOnes(PtrTy.getSizeInBits()), DL,
+                            PtrTy)));
+
+  if (VT != MVT::v16i1)
```
The hard coded i16 / v16i1 makes me nervous :)

abhishek-kaushik22 replied:
From my understanding you can't really extract more than 16 bits using this pattern, because $$2^{17}$$ will not fit in `i16` and a `vXi32` where $$X > 16$$ is more than 512 bits.

RKSimon replied:
Add a comment to explain this

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -55447,6 +55447,81 @@ static SDValue truncateAVX512SetCCNoBWI(EVT VT, EVT OpVT, SDValue LHS,
+      UndefElts[0] || !EltBits[0].isPowerOf2())
+    return SDValue();
+
+  unsigned N = EltBits[0].logBase2();
```
Code comment explaining what you're testing.

abhishek-kaushik22 commented:
> @abhishek-kaushik22 Can you confirm if this will fix #72803 please?

Yes, for 
```llvm
define void @example(ptr noalias nocapture noundef writeonly sret([8 x i32]) align 4 dereferenceable(32) %_0, ptr noalias nocapture noundef readonly align 4 dereferenceable(32) %a, i64 noundef %m, ptr noalias nocapture noundef readonly align 4 dereferenceable(32) %b) {
start:
  %0 = insertelement <8 x i64> poison, i64 %m, i64 0
  %1 = shufflevector <8 x i64> %0, <8 x i64> poison, <8 x i32> zeroinitializer
  %2 = and <8 x i64> %1, <i64 1, i64 2, i64 4, i64 8, i64 16, i64 32, i64 64, i64 128>
  %3 = icmp eq <8 x i64> %2, zeroinitializer
  %4 = load <8 x i32>, ptr %b, align 4
  %5 = select <8 x i1> %3, <8 x i32> zeroinitializer, <8 x i32> %4
  %6 = load <8 x i32>, ptr %a, align 4
  %7 = add <8 x i32> %6, %5
  store <8 x i32> %7, ptr %_0, align 4
  ret void
}
```
this generates 
```asm
        movq	%rdi, %rax
	kmovd	%edx, %k1
	vmovdqu	(%rsi), %ymm0
	vpaddd	(%rcx), %ymm0, %ymm0 {%k1}
	vmovdqu	%ymm0, (%rdi)
	vzeroupper
	retq
```

Commit:
fix reviews

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -55447,6 +55447,85 @@ static SDValue truncateAVX512SetCCNoBWI(EVT VT, EVT OpVT, SDValue LHS,
+
+  SDValue BroadcastOp = Broadcast.getOpcode() == X86ISD::VBROADCAST
+                            ? Broadcast.getOperand(0)
+                            : Broadcast.getOperand(1);
```
This won't work for VBROADCAST_LOAD - you're better off just using a EXTRACT_VECTOR_ELT node to extract the 0th element from Broadcast and let existing folds handle it.

abhishek-kaushik22 replied:
Weirdly if I do it like that, the `vpbroadcastb` is still there in the assembly followed by a `vmovd`

RKSimon replied:
OK, just for BROADCAST_LOAD then

RKSimon commented on llvm/test/CodeGen/X86/kmov.ll:
```diff
@@ -0,0 +1,372 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc < %s -mtriple=x86_64-- -mcpu=x86-64-v4 | FileCheck %s
```
Add a -mcpu=knl run to check what happens on AVX512 targets with less ISA support

abhishek-kaushik22 replied:
> Add a -mcpu=knl run to check what happens on AVX512 targets with less ISA support

This one doesn't really work as expected for most of the cases because of the `PCMPEQ` node.

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -55579,6 +55658,12 @@ static SDValue combineSetCC(SDNode *N, SelectionDAG &DAG,
+      if (SDValue V =
+              combineAVX512SetCCToKMOV(VT, Op0, TmpCC, DL, DAG, Subtarget))
+        return V;
+    }
```
(style) remove outer braces

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -55447,6 +55447,85 @@ static SDValue truncateAVX512SetCCNoBWI(EVT VT, EVT OpVT, SDValue LHS,
+      break;
+    }
+
+    if (EltBits[I] != 1ULL << (N + I))
```
`EltBits[I].isOneBitSet(N)` ?

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -55447,6 +55447,85 @@ static SDValue truncateAVX512SetCCNoBWI(EVT VT, EVT OpVT, SDValue LHS,
+        ISD::XOR, DL, MVT::v16i1, Bitcast,
+        DAG.getSplatBuildVector(
+            MVT::v16i1, DL,
+            DAG.getConstant(APInt::getAllOnes(PtrTy.getSizeInBits()), DL,
```
`DAG.getAllOnesConstant`

Commit:
fix reviews

Commit:
Update X86ISelLowering.cpp

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -55447,6 +55447,93 @@ static SDValue truncateAVX512SetCCNoBWI(EVT VT, EVT OpVT, SDValue LHS,
+    BroadcastOp = DAG.getNode(ISD::EXTRACT_VECTOR_ELT, DL, BroadcastOpVT,
+                              Broadcast, DAG.getConstant(0, DL, VecIdxTy));
+  } else {
+    BroadcastOp = Broadcast.getOperand(0);
```
This still assumes BroadcastOp is a scalar - its valid for X86ISD::VBROADCAST to take a vector

abhishek-kaushik22 replied:
I've added a new check so that we only combine scalar inputs to `VBROADCAST`

Commit:
Update X86ISelLowering.cpp

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -55447,6 +55447,95 @@ static SDValue truncateAVX512SetCCNoBWI(EVT VT, EVT OpVT, SDValue LHS,
+
+  SDValue Masked = BroadcastOp;
+  if (N != 0) {
+    unsigned Mask = (1ULL << Len) - 1;
```
Better to create Mask with APInt::getLowBitsSet

RKSimon reviewed (commented):
One minor, but I think its almost there

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -55447,6 +55447,95 @@ static SDValue truncateAVX512SetCCNoBWI(EVT VT, EVT OpVT, SDValue LHS,
+                         DAG.getConstant(Mask, DL, BroadcastOpVT));
+  }
+  // We can't extract more than 16 bits using this pattern, because 2^{17} will
+  // not fit in an i16 and a vXi32 where X > 16 is more than 512 bits.
```
Where is the check that VT is not greater than v16i1?

abhishek-kaushik22 replied:
I've added the check here: `UndefElts.getBitWidth() > 16`

Commit:
Address review comments

Commit:
Merge branch 'main' into kmov

Commit:
Fix tests

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -55688,6 +55688,95 @@ static SDValue truncateAVX512SetCCNoBWI(EVT VT, EVT OpVT, SDValue LHS,
+    return SDValue();
+
+  SDValue Load = Op0.getOperand(1);
+  EVT LoadVT = Load.getSimpleValueType();
```
Remove LoadVT - this should match VT

abhishek-kaushik22 replied:
Sorry, `VT` is the type of `setcc` (something like `v*i1`) and doesn't match the type of load (which is `v*i64`)

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -55688,6 +55688,95 @@ static SDValue truncateAVX512SetCCNoBWI(EVT VT, EVT OpVT, SDValue LHS,
+
+  APInt UndefElts;
+  SmallVector<APInt, 32> EltBits;
+  if (!getTargetConstantBitsFromNode(Load, LoadVT.getScalarSizeInBits(),
```
Use Op0.getOperand(1) instead of Load - and remove the Load variable

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -55688,6 +55688,95 @@ static SDValue truncateAVX512SetCCNoBWI(EVT VT, EVT OpVT, SDValue LHS,
+  SDValue BroadcastOp;
+  if (Broadcast.getOpcode() != X86ISD::VBROADCAST) {
+    BroadcastOp = DAG.getNode(ISD::EXTRACT_VECTOR_ELT, DL, BroadcastOpVT,
+                              Broadcast, DAG.getConstant(0, DL, VecIdxTy));
```
Replace getConstant with getVectorIdxConstant (and remove DataLayout/VecIdxTy).

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -55688,6 +55688,95 @@ static SDValue truncateAVX512SetCCNoBWI(EVT VT, EVT OpVT, SDValue LHS,
+
+  if (VT != MVT::v16i1)
+    return DAG.getNode(ISD::EXTRACT_SUBVECTOR, DL, VT, Bitcast,
+                       DAG.getConstant(0, DL, PtrTy));
```
getConstant -> getVectorIdxConstant

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -55688,6 +55688,95 @@ static SDValue truncateAVX512SetCCNoBWI(EVT VT, EVT OpVT, SDValue LHS,
+  // not fit in an i16 and a vXi32 where X > 16 is more than 512 bits.
+  SDValue Trunc = DAG.getAnyExtOrTrunc(Masked, DL, MVT::i16);
+  SDValue Bitcast = DAG.getNode(ISD::BITCAST, DL, MVT::v16i1, Trunc);
+  MVT PtrTy = TLI.getPointerTy(DataLayout);
```
remove

abhishek-kaushik22 replied:
This is used in `DAG.getAllOnesConstant(DL, PtrTy)`

RKSimon reviewed (commented):
A few final minors

Commit:
Remove basic block name from tests

Commit:
Use getVectorIdxConstant instead of getConstant

Commit:
Merge branch 'main' into kmov

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -55689,6 +55689,94 @@ static SDValue truncateAVX512SetCCNoBWI(EVT VT, EVT OpVT, SDValue LHS,
+
+  if (CC == ISD::SETEQ)
+    Bitcast =
+        DAG.getNode(ISD::XOR, DL, MVT::v16i1, Bitcast,
```
Can't this be handled by DAG.getNOT() ?

abhishek-kaushik22 replied:
Thanks! I have removed all the unnecessary declarations.

Commit:
Use DAG.getNOT

RKSimon reviewed (approved):
LGTM - cheers

abhishek-kaushik22 commented:
Thank You @RKSimon :)

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -55689,6 +55689,88 @@ static SDValue truncateAVX512SetCCNoBWI(EVT VT, EVT OpVT, SDValue LHS,
+
+  SDValue Masked = BroadcastOp;
+  if (N != 0) {
+    APInt Mask = APInt::getLowBitsSet(BroadcastOpVT.getSizeInBits(), Len);
```
@abhishek-kaushik22 what happens for <16 x i8> cases where Len (== 16) > BroadcastOpVT.getSizeInBits() (== 8)? I think I have a regression case but its taking forever to reduce.

abhishek-kaushik22 replied:
Thanks! I can see that would be a problem. I am trying to get a test case for this.
